### ISSUE TITLE
sql: panic: expected *DString, found *tree.DTuple

### ISSUE THREAD

**Issue by mjibson** [cockroachdb/cockroach#40440]
```
CREATE TABLE IF NOT EXISTS tab_orig AS
	SELECT
		g::STRING AS _string,
		substring(NULL, NULL, NULL)::UUID AS _uuid
	FROM
		generate_series(0, 0) AS g;

WITH
	with_669 AS (SELECT NULL)
SELECT
	NULL
FROM
	tab_orig AS tab_3129
WHERE
	NOT
		EXISTS(
			WITH
				with_670
					AS (
						SELECT
							*
						FROM
							(
								VALUES
									(
										COALESCE(
											CASE
											WHEN false
											THEN 0
											ELSE NULL
											END,
											0
										),
										"right"(
											convert_to(
												quote_nullable(
													NULL
												),
												rtrim(
													(
														SELECT
															tab_3130._string
														FROM
															tab_orig
																AS tab_3130
															JOIN (
																	VALUES
																		(
																			''
																		)
																)
																	AS tab_3131
																		(
																			col_8159
																		) ON
																	tab_3130._string
																	= tab_3131.col_8159
													),
													''
												)
											),
											0
										),
										0:::OID,
										'44eb:6869:fe3b:8745:e190:661f:fd38:5d1f/107':::INET,
										''
									)
							)
								AS tab_3133
									(
										col_8161,
										col_8162,
										col_8163,
										col_8164,
										col_8165
									)
						EXCEPT
							SELECT
								*
							FROM
								(
									VALUES
										(
											0,
											'':::BYTES,
											0:::OID,
											'170.166/12':::INET,
											''
										)
								)
									AS tab_3134
										(
											col_8166,
											col_8167,
											col_8168,
											col_8169,
											col_8170
										)
					)
			SELECT
				tab_3129._uuid
		);
```

```
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc001179000, 0x4d2e1a0, 0xc000a270c0, 0x4043ee0, 0xc0004266f0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:730 +0x330
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1(0xc001179000, 0x4d2e1a0, 0xc000a270c0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:442 +0x61
panic(0x4043ee0, 0xc0004266f0)
	/usr/local/go/src/runtime/panic.go:522 +0x1b5
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.MustBeDString(0x4d52060, 0xc0018052c0, 0xc0018052c0, 0x4d52060)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/datum.go:1098 +0xe3
github.com/cockroachdb/cockroach/pkg/sql/sem/builtins.stringOverload2.func1(0xc00113d440, 0xc0003de300, 0x2, 0x2, 0x2, 0x2, 0x0, 0x0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/builtins/builtins.go:3905 +0x79
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*FuncExpr).Eval(0xc001389b80, 0xc00113d440, 0xc001389b80, 0x4d7c1a0, 0xc001389b80, 0x0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/eval.go:3832 +0xb8
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*FuncExpr).evalArgs(0xc001389c00, 0xc00113d440, 0x38, 0x7f5da4078d00, 0x7f5da41eefff, 0xc001597320, 0x711ede, 0x7f5da4078b00)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/eval.go:3810 +0xca
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*FuncExpr).Eval(0xc001389c00, 0xc00113d440, 0xc001389c00, 0x4d7c1a0, 0xc001389c00, 0x0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/eval.go:3824 +0x50
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*FuncExpr).evalArgs(0xc001389c80, 0xc00113d440, 0x1, 0xc001136000, 0xc001597540, 0x1ea3be4, 0x4127a40, 0x140)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/eval.go:3810 +0xca
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*FuncExpr).Eval(0xc001389c80, 0xc00113d440, 0x4db3e80, 0xc0001803a8, 0x0, 0x0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/eval.go:3824 +0x50
github.com/cockroachdb/cockroach/pkg/sql.(*valuesNode).startExec(0xc001843630, 0x4d2e1a0, 0xc000a26bc0, 0xc00113d440, 0xc001179358, 0x1c3d5b6, 0xc001597778)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/values.go:156 +0x366
github.com/cockroachdb/cockroach/pkg/sql.startExec.func2(0x438c53c, 0x6, 0x4d30660, 0xc001843630, 0xc000f7a660, 0xc001597788)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:362 +0x55
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal.func1(0xc00166e8a0, 0x438c53c, 0x6, 0x4d30660, 0xc001843630)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:141 +0x5d
```

**Comment by jordanlewis** [comment]
Extra reductions that would have helped here:

1. replace case expression with first arm of case expression
2. replace `coalesce(x)` with `x`

**Comment by jordanlewis** [comment]
The `explain(types)` produces a bunch of "unknown"-typed subqueries. This seems problematic.

```
             tree            |     field     |                                                                                                                                                                                                                                                     description                                                                                                                                                                                                                                                     |                                 columns                                 | ordering
+----------------------------+---------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------------------------------------------------------------------------+----------+
  root                       |               |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     | ("?column?" unknown)                                                    |
   ├── render                |               |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     | ("?column?" unknown)                                                    |
   │    │                    | render 0      | (NULL)[unknown]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     |                                                                         |
   │    └── apply-join       |               |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     | (_uuid uuid)                                                            |
   │         │               | type          | anti                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                |                                                                         |
   │         └── scan        |               |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     | (_uuid uuid)                                                            |
   │                         | table         | tab_orig@primary                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |                                                                         |
   │                         | spans         | ALL                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 |                                                                         |
   ├── subquery              |               |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     | ("?column?" unknown)                                                    |
   │    │                    | id            | @S1                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 |                                                                         |
   │    │                    | original sql  | SELECT NULL                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         |                                                                         |
   │    │                    | exec mode     | all rows                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            |                                                                         |
   │    └── buffer node      |               |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     | ("?column?" unknown)                                                    |
   │         │               | label         | buffer 1 (with_669)                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 |                                                                         |
   │         └── values      |               |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     | ("?column?" unknown)                                                    |
   │                         | size          | 1 column, 1 row                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     |                                                                         |
   │                         | row 0, expr 0 | (NULL)[unknown]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     |                                                                         |
   ├── subquery              |               |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     | ("?column?" unknown)                                                    |
   │    │                    | id            | @S2                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 |                                                                         |
   │    │                    | original sql  | (SELECT tab_3130._string FROM tab_orig AS tab_3130 JOIN (VALUES ('')) AS tab_3131 (col_8159) ON tab_3130._string = tab_3131.col_8159)                                                                                                                                                                                                                                                                                                                                                                               |                                                                         |
   │    │                    | exec mode     | one row                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             |                                                                         |
   │    └── max1row          |               |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     | (_string string)                                                        |
   │         └── scan        |               |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     | (_string string)                                                        |
   │                         | table         | tab_orig@primary                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |                                                                         |
   │                         | spans         | ALL                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 |                                                                         |
   │                         | filter        | ((_string)[string] = ('')[string])[bool]                                                                                                                                                                                                                                                                                                                                                                                                                                                                            |                                                                         |
   └── subquery              |               |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     | ("?column?" unknown)                                                    |
        │                    | id            | @S3                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 |                                                                         |
        │                    | original sql  | SELECT * FROM (VALUES (COALESCE(0), right(convert_to(quote_nullable(NULL), rtrim((SELECT tab_3130._string FROM tab_orig AS tab_3130 JOIN (VALUES ('')) AS tab_3131 (col_8159) ON tab_3130._string = tab_3131.col_8159), '')), 0), 0:::OID, '44eb:6869:fe3b:8745:e190:661f:fd38:5d1f/107':::INET, '')) AS tab_3133 (col_8161, col_8162, col_8163, col_8164, col_8165) EXCEPT SELECT * FROM (VALUES (0, '':::BYTES, 0:::OID, '170.166/12':::INET, '')) AS tab_3134 (col_8166, col_8167, col_8168, col_8169, col_8170) |                                                                         |
        │                    | exec mode     | all rows                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            |                                                                         |
        └── buffer node      |               |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     | (column1 int, column2 bytes, column3 oid, column4 inet, column5 string) |
             │               | label         | buffer 2 (with_670)                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 |                                                                         |
             └── union       |               |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     | (column1 int, column2 bytes, column3 oid, column4 inet, column5 string) |
                  ├── values |               |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     | (column1 int, column2 bytes, column3 oid, column4 inet, column5 string) |
                  │          | size          | 5 columns, 1 row                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |                                                                         |
                  │          | row 0, expr 0 | (0)[int]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            |                                                                         |
                  │          | row 0, expr 1 | (right((convert_to(('NULL')[string], (rtrim((@S2)[string], ('')[string]))[string]))[bytes], (0)[int]))[bytes]                                                                                                                                                                                                                                                                                                                                                                                                       |                                                                         |
                  │          | row 0, expr 2 | (0)[oid]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            |                                                                         |
                  │          | row 0, expr 3 | ('44eb:6869:fe3b:8745:e190:661f:fd38:5d1f/107')[inet]                                                                                                                                                                                                                                                                                                                                                                                                                                                               |                                                                         |
                  │          | row 0, expr 4 | ('')[string]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        |                                                                         |
                  └── values |               |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     | (column1 int, column2 bytes, column3 oid, column4 inet, column5 string) |
                             | size          | 5 columns, 1 row                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |                                                                         |
                             | row 0, expr 0 | (0)[int]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            |                                                                         |
                             | row 0, expr 1 | ('\x')[bytes]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       |                                                                         |
                             | row 0, expr 2 | (0)[oid]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            |                                                                         |
                             | row 0, expr 3 | ('170.166.0.0/12')[inet]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            |                                                                         |
                             | row 0, expr 4 | ('')[string]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        |                                                                         |
(47 rows)
```

**Comment by jordanlewis** [comment]
`explain(opt,verbose)`:

```
                                                                             text
+------------------------------------------------------------------------------------------------------------------------------------------------------------+
  with &1 (with_669)
   ├── columns: "?column?":20
   ├── stats: [rows=1]
   ├── cost: 2.45
   ├── fd: ()-->(20)
   ├── values
   │    ├── columns: "?column?":1
   │    ├── cardinality: [1 - 1]
   │    ├── stats: [rows=1]
   │    ├── cost: 0.02
   │    ├── key: ()
   │    ├── fd: ()-->(1)
   │    └── (NULL,)
   └── project
        ├── columns: "?column?":20
        ├── stats: [rows=1]
        ├── cost: 2.42
        ├── fd: ()-->(20)
        ├── anti-join-apply
        │    ├── columns: tab_3129._uuid:3
        │    ├── stats: [rows=1]
        │    ├── cost: 2.39
        │    ├── scan tab_3129
        │    │    ├── columns: tab_3129._uuid:3
        │    │    ├── stats: [rows=1]
        │    │    ├── cost: 1.1
        │    │    └── prune: (3)
        │    ├── with &2 (with_670)
        │    │    ├── columns: "_uuid":19
        │    │    ├── outer: (3)
        │    │    ├── cardinality: [1 - 1]
        │    │    ├── stats: [rows=1]
        │    │    ├── cost: 1.24
        │    │    ├── key: ()
        │    │    ├── fd: ()-->(19)
        │    │    ├── except
        │    │    │    ├── columns: column1:9 column2:10 column3:11 column4:12 column5:13
        │    │    │    ├── left columns: column1:9 column2:10 column3:11 column4:12 column5:13
        │    │    │    ├── right columns: column1:14 column2:15 column3:16 column4:17 column5:18
        │    │    │    ├── cardinality: [0 - 1]
        │    │    │    ├── stats: [rows=1, distinct(9-13)=1, null(9-13)=0]
        │    │    │    ├── cost: 1.21
        │    │    │    ├── key: (9-13)
        │    │    │    ├── values
        │    │    │    │    ├── columns: column1:9 column2:10 column3:11 column4:12 column5:13
        │    │    │    │    ├── cardinality: [1 - 1]
        │    │    │    │    ├── stats: [rows=1, distinct(9-13)=1, null(9-13)=0]
        │    │    │    │    ├── cost: 1.15
        │    │    │    │    ├── key: ()
        │    │    │    │    ├── fd: ()-->(9-13)
        │    │    │    │    └── tuple
        │    │    │    │         ├── const: 0
        │    │    │    │         ├── function: right
        │    │    │    │         │    ├── function: convert_to
        │    │    │    │         │    │    ├── quote_nullable(NULL)
        │    │    │    │         │    │    └── function: rtrim
        │    │    │    │         │    │         ├── subquery
        │    │    │    │         │    │         │    └── max1-row
        │    │    │    │         │    │         │         ├── columns: tab_3130._string:5
        │    │    │    │         │    │         │         ├── cardinality: [0 - 1]
        │    │    │    │         │    │         │         ├── stats: [rows=1]
        │    │    │    │         │    │         │         ├── cost: 1.13
        │    │    │    │         │    │         │         ├── key: ()
        │    │    │    │         │    │         │         ├── fd: ()-->(5)
        │    │    │    │         │    │         │         └── select
        │    │    │    │         │    │         │              ├── columns: tab_3130._string:5
        │    │    │    │         │    │         │              ├── stats: [rows=1, distinct(5)=1, null(5)=0]
        │    │    │    │         │    │         │              ├── cost: 1.12
        │    │    │    │         │    │         │              ├── fd: ()-->(5)
        │    │    │    │         │    │         │              ├── scan tab_3130
        │    │    │    │         │    │         │              │    ├── columns: tab_3130._string:5
        │    │    │    │         │    │         │              │    ├── stats: [rows=1, distinct(5)=1, null(5)=0]
        │    │    │    │         │    │         │              │    ├── cost: 1.1
        │    │    │    │         │    │         │              │    └── prune: (5)
        │    │    │    │         │    │         │              └── filters
        │    │    │    │         │    │         │                   └── tab_3130._string = '' [outer=(5), constraints=(/5: [/'' - /'']; tight), fd=()-->(5)]
        │    │    │    │         │    │         └── const: ''
        │    │    │    │         │    └── const: 0
        │    │    │    │         ├── const: 0
        │    │    │    │         ├── const: '44eb:6869:fe3b:8745:e190:661f:fd38:5d1f/107'
        │    │    │    │         └── const: ''
        │    │    │    └── values
        │    │    │         ├── columns: column1:14 column2:15 column3:16 column4:17 column5:18
        │    │    │         ├── cardinality: [1 - 1]
        │    │    │         ├── stats: [rows=1, distinct(14-18)=1, null(14-18)=0]
        │    │    │         ├── cost: 0.02
        │    │    │         ├── key: ()
        │    │    │         ├── fd: ()-->(14-18)
        │    │    │         └── (0, '\x', 0, '170.166.0.0/12', '')
        │    │    └── values
        │    │         ├── columns: "_uuid":19
        │    │         ├── outer: (3)
        │    │         ├── cardinality: [1 - 1]
        │    │         ├── stats: [rows=1]
        │    │         ├── cost: 0.02
        │    │         ├── key: ()
        │    │         ├── fd: ()-->(19)
        │    │         └── (tab_3129._uuid,)
        │    └── filters (true)
        └── projections
             └── null
(101 rows)
```

**Comment by mjibson** [comment]
This query is now giving:

```
Error: pq: internal error: estimated row count must be non-zero
DETAIL: stack trace:
github.com/cockroachdb/cockroach/pkg/sql/opt/memo/statistics_builder.go:2279: finalizeFromCardinality()
github.com/cockroachdb/cockroach/pkg/sql/opt/memo/statistics_builder.go:735: buildProject()
github.com/cockroachdb/cockroach/pkg/sql/opt/memo/logical_props_builder.go:323: buildProjectProps()
github.com/cockroachdb/cockroach/pkg/sql/opt/memo/expr.og.go:15154: MemoizeProject()
github.com/cockroachdb/cockroach/pkg/sql/opt/norm/factory.og.go:1555: ConstructProject()
github.com/cockroachdb/cockroach/pkg/sql/opt/norm/factory.og.go:1284: ConstructProject()
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/project.go:58: constructProject()
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/project.go:30: constructProjectForScope()
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/select.go:851: buildSelectClause()
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/select.go:757: buildSelect()
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/builder.go:241: buildStmt()
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/builder.go:185: Build()
github.com/cockroachdb/cockroach/pkg/sql/plan_opt.go:419: buildExecMemo()
github.com/cockroachdb/cockroach/pkg/sql/plan_opt.go:161: makeOptimizerPlan()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:734: makeExecPlan()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:616: dispatchToExecutionEngine()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:417: execStmtInOpenState()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:98: execStmt()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1238: execCmd()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1167: run()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:444: ServeConn()
github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:584: func1()
runtime/asm_amd64.s:1337: goexit()
```

**Comment by rytaft** [comment]
This issue appears to be fixed by #40472. The query executes successfully with that PR. Once it merges, should I mark this as fixed? Or is the original issue still a problem?

**Comment by rytaft** [comment]
With the fix from #40472, the plan is reduced to:
```
  values                                     
   ├── columns: "?column?":20(unknown!null)  
   ├── cardinality: [0 - 0]                  
   ├── stats: [rows=0]                       
   ├── cost: 0.01                            
   ├── key: ()                               
   ├── fd: ()-->(20)                         
   └── prune: (20)                           
```

**Comment by mjibson** [comment]
Yes, we can mark this as fixed by that.
